Refactor encode / decode functions to use `bytes::Buf` and `bytes::BufMut`

xoac opened the pull request:
Changes are listed in CHANGELOG.md

I want to make this crate more handy for people that can use allocation but it should be fully compatible with manually allocated buffers (as it was before this PR)

I couldn't done this for `Frame<'p>::encode()` since I couldn't force it to work with `& dyn Buf`. More details are in in [this issue](https://github.com/tokio-rs/bytes/issues/395)

Closes #9

xoac commented:
One thing that this PR doesn't contains  is how to use `BufMut` and that it may panic. It's documented already in [`BufMut::advance()`](https://docs.rs/bytes/0.5.4/bytes/trait.BufMut.html#panics)

I am not native and I don't fell good in writing such documentation. Before I star could you let me know this changes will be accepted?

hannobraun commented:
Thanks you for the pull request, @xoac!

Unfortunately, I don't have a lot of time to review this right now, but I'll try to do so as soon as possible.

xoac commented:
Thanks.
There are a lot of lines changed but this is mostly mechanics. Here is a list of conceptually changes:
- `enode(buf: &mut [u8])` -> `encode(buf: &mut dyn BufMut)` 
- `decode(buf: &[u8]`) -> `decode(buf: &mut dyn Buf)` (Frame::deocde is exception - linked issue).
- `Address::None` was removed and replaced with `Option<Address>`
- Header was modified a little to match exactly documentation:
  - added `FrameControl` struct and moved logic and related struct to separate file.

Functionality there is no changes in this PR. Just resolving #9. And making implementation simpler.
Bugs could appear when I changed from `buf.len()` to `buf.remaining()`. 
Now it should be easier to implement new things since there is no need to count bytes anymore.

hannobraun commented:
Thanks for the summary, @xoac. It looks good to me. I had a short look already, and what I saw looks good to me also.

Now I just need to find the time to do a proper review, hopefully soon.

hannobraun commented on src/mac/frame.rs:
```diff
@@ -1,1209 +0,0 @@
-/// Part of [`Header`]. Auxiliary security headers are currently unsupported by
-/// this implementation.
-#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
-pub enum Security {
```
I don't like the removal of the `Security` enum. It was added to clearly document that auxiliary security headers are not supported currently. Without it, someone might set `security` to `true` and wonder why it doesn't work.

xoac replied:
Agree. But it would be better to leave bool and implement security. But until it's unimplemented I will restore this structure.

hannobraun replied:
Sounds good!

hannobraun commented on src/mac/frame/frame_control.rs:
```diff
@@ -0,0 +1,262 @@
+    pub pan_id_compress: bool,
+
+    /// Destination address mode
+    pub dest_addr_mode: AddressMode,
```
The information in this field is redundant with the fields containing `Address`. What happens if the user creates a `Frame` in with a short destination address, but adds `AddressMode::Extended` here? What should happen in such a case?

xoac replied:
Nice catch, I think in this case we need builder. At there should be at least `debug_assert!` similar to [this one](https://github.com/xoac/rust-ieee802.15.4/blob/19c986705b177ac4dc7114550f123d6d3da2b6e8/src/mac/frame/mod.rs#L399)

Having this redundant is nice when you parsing frame. This make this structure looks exactly like in IEEE Std 802.15.4 what adds readability. Also functions are shorter more local.

The second solution is to remove FrameControl and make it fields part of Header (I think as it was before). 

In my opinion the second solution is better.

hannobraun replied:
Yes, I fully agree. Builder methods and making the fields private would solve this, but restoring the previous situation (`FrameControl` fields being part of `Header`) would be the much simpler solution.

hannobraun commented on src/mac/frame/frame_control.rs:
```diff
@@ -0,0 +1,262 @@
+    pub version: FrameVersion,
+
+    /// Source address mode
+    pub src_addr_mode: AddressMode,
```
Same problem as with `dest_addr_mode`.

hannobraun commented on src/mac/frame/mod.rs:
````diff
@@ -0,0 +1,993 @@
+    /// ];
+    /// assert_eq!(encoded_bytes[..], expected_bytes[..]);
+    /// ```
+    pub fn encode(&self, buf: &mut dyn BufMut, write_footer: WriteFooter) {
````
I might be missing something, but I think that this no longer returns the length of the encoded frame makes this very inconvenient to use with a `&mut [u8]`, if you need to know that length. As far as I can see it, you'd need to do something like this:
``` rust
let len_before = slice.len();
frame.encode(&mut slice);
let len = len_before - slice.len();
```

Is there an easier way? If not, can we return the length of the encoded frame here?

xoac replied:
In my opinion we should not enhance user to use raw `&mut [u8]` because  it can panic - (so it's not so rusty). I would prefer to wrap `&mut [u8]` in some wrapper and implement with `BufMut`:
- flag that validate slice end was reached
- allow determinate proper lengths (encoded/free)

I thought such crate should already exist - but could not find one. I could implement this and add it to example.
The second advantage with this implementation will be speed - if you are 100% sure you have enough space in slice - you don't need calculate byte counter. (I assume we want to return how many bytes was encoded from all `encode()` functions not just `Frame::encode()`)  This crate handle mac frames - so this can matter. 

If you need to call it often you can always create a helper function. 
```
fn encode_n(frame: &Frame, mut slice &mut[u8], write_fotter: WriteFooter) -> usize {
  let len_before = slice.len();
  frame.encode(&mut slice);
  return len_before - slice.len();
}
```

hannobraun replied:
The reason for my comment above is that I ported my [downstream crate](https://github.com/braun-embedded/rust-dw1000/tree/master/dw1000) to your branch and the code that calls `encode`, which had been straight-forward before, not only became more complicated, but it became hard to figure out what needed to be done in the first place.

I had to study `BufMut`, including its code, to figure out that it updates the slice in-place, and I can use `slice.len()` to figure out how much was written. This was fine for me (I had to understand `BufMut` anyway, for this review), but I think it puts too much of a burden on the user, especially since that was so easy before this change.

I think there are two solutions:
1. Just return the length, as I suggested. You make a good point about the panicking, but I'd be willing to live with that, as that's pretty much the same as we had previously.
2. We add a wrapper, as you suggested. I think this would be the better solution, and if you're willing to do that, that's great! If you don't have the time or inclination, I think solution 1. is a good-enough stopgap.

> (I assume we want to return how many bytes was encoded from all `encode()` functions not just `Frame::encode()`)

My downstream crate only uses the top-level `encode()`, which I assume is a fairly typical use of this crate. So this one is the most important. But handling it consistently for all `encode()` methods would be the best solution.

xoac replied:
I don't see ability to adding returning length just for `Frame::encode()` or I don't know how to do it. So change has to be done to all `encode()` what I would be doing anyway -- interface should be  the same for all functions.

> I had to study BufMut, including its code, to figure out that it updates the slice in-place, and I can use slice.len() to figure out how much was written. This was fine for me (I had to understand BufMut anyway, for this review), but I think it puts too much of a burden on the user, especially since that was so easy before this change.

I think we want to resolve #9. This is proposed cost of this - you need to study why `&mut [u8]` panics and should workaround. We can of course show in an example how to do it "proper".

I have found that there is [heapless](https://crates.io/crates/heapless) crate. Maybe it would be good way to add there PR that support `bytes::BufMut`.

I will have time to work on this after 24.08.

hannobraun replied:
Sounds good!

xoac replied:
Could you please test if [static-bytes](https://github.com/xoac/static-bytes) crate would be more user-friendly than `&mut [u8]` slice in your case?

I have created [one test](https://github.com/xoac/static-bytes/blob/290c1c366f94c71c1b75cad98d5a0981bb477a74/src/bytes_slice.rs#L104) that compares the `&mut [u8]` and `SafeBytesSlice`. It can be a good place to see how to use SafeBytesSlice

hannobraun commented on src/mac/mod.rs:
```diff
@@ -4,9 +4,4 @@ pub mod beacon;
-    WriteFooter,
-};
-
-pub use frame::frame_control::{AddressMode, FrameControl, FrameType, FrameVersion};
```
I'd prefer if all the important types still were re-exported. Without the re-exports, using the library becomes less convenient, and I don't think the internal structure of the module is important to the user.

This is also a breaking change, which I'd like to avoid unless there is a clear benefit.

hannobraun reviewed (changes requested):
Thank you, @xoac, `static-bytes` is exactly what's needed here!

I've found one more problem in the last commit (left a comment), but otherwise I'm happy. Once you re-add the re-exports, this is ready to merge, I think.

Of course, more documentation (especially mentioning `static-bytes`) would always be nice, but I'm happy to merge this PR without it, if you prefer.

hannobraun reviewed (approved):
This look good now. Thanks again @xoac, for making this change, and also thank you for being persistent in getting this ready to merge.

Oh, and thank you for making `static-bytes`!
